### PR TITLE
Logging for WASM httpCall

### PR DESCRIPTION
**Pull request by esnible**
This is a recreation of https://github.com/envoyproxy/envoy-wasm/pull/640

I had some difficulties troubleshooting a WASM filter that made an httpCall. I needed some simple logging to see what my AssemblyScript was doing incorrectly. This is my first contribution to this repo.

Commit Message: Logging for WASM httpCall
Additional Description: When writing filters in AssemblyScript it is easy to make coding mistakes. Envoy will return WasmResult::BadArgument but it isn't easy to know which argument is the problem. My code had two mistakes and these log messages helped me find them.
Risk Level: Low
Testing: I tested this manually applied to envoyproxy/envoy-wasm, in Istio, building a Docker image using https://github.com/istio/proxy containing an Envoy with these changes and then manually inspected the logs while troubleshooting my WASM filter.
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by esnible** [comment]
@PiotrSikora I am looking for advice.

- I am on a Mac.  I am nervous about enabling the Git hooks suggested by the Envoy docs.
- How can I run just the linux_x64 coverage locally manually?
- I suspect code coverage tests are failing because either the tests aren't logging at a detailed-enough level to trigger my trace logging, or because my debug logging isn't firing because no tests pass invalid values to WASM `httpGet()`.  How do I see detailed info on why the code coverage failed?

**Comment by PiotrSikora** [comment]
> @PiotrSikora I am looking for advice.
> 
> * I am on a Mac.  I am nervous about enabling the Git hooks suggested by the Envoy docs.
> * How can I run just the linux_x64 coverage locally manually?
> * I suspect code coverage tests are failing because either the tests aren't logging at a detailed-enough level to trigger my trace logging, or because my debug logging isn't firing because no tests pass invalid values to WASM `httpGet()`.  How do I see detailed info on why the code coverage failed?

I don't have those Git hooks enabled, they are not mandatory, so if you're nervous, then you can just skip them.

Coverage failure is unrelated to your changes. Other PRs are also failing because of it, and the error on the CI says: `Code coverage for source/server is lower than limit of 94.6 (94.5)`, so you need wait until it's fixed on `master` branch, and then update your PR.

**Comment by esnible** [comment]
@PiotrSikora what would it take to get a review?  This has been sitting here for two weeks, now it has a conflict.  Should I rebase and wait?

This inspiration for this was the difficulties I had using WASM with AssemblyScript to write my own plugin.  It can be extremely difficult to debug a plugin without logging.  For example, I wanted to invoke an httpCall() in the onResponse() callback instead of the onRequest() callback.  It turns out the Solo.io runtime forgets the request headers.  I had printed them earlier, I passed them to httpCall().  It would have been super handy to see a log statement telling me what was wrong.  I spent several days trying to figure this out.

This change seems to have minimal risk.  It only affects people using WASM with Envoy running with DEBUG logging.

File: source/extensions/common/wasm/context.cc [hunk 1]
```diff
@@ -146,7 +146,10 @@ Context::Context(Wasm* wasm, const PluginSharedPtr& plugin) : ContextBase(wasm,
   root_local_info_ = &std::static_pointer_cast<Plugin>(plugin)->local_info_;
 }
 Context::Context(Wasm* wasm, uint32_t root_context_id, const PluginSharedPtr& plugin)
-    : ContextBase(wasm, root_context_id, plugin) {}
+    : ContextBase(wasm, root_context_id, plugin) {
+  ENVOY_LOG(trace, "wasm log{}: Context::Context(..., root_context_id={}, ...)", log_prefix(),
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/context.cc:150]: `wasm log` prefix is used for logs generated from within WasmVM, so adding it for things logged from outside the WasmVM is a bit confusing.

File: source/extensions/common/wasm/context.cc [hunk 2]
```diff
@@ -889,10 +893,13 @@ WasmResult Context::httpCall(absl::string_view cluster, const Pairs& request_hea
                              absl::string_view request_body, const Pairs& request_trailers,
                              int timeout_milliseconds, uint32_t* token_ptr) {
   if (timeout_milliseconds < 0) {
+    ENVOY_LOG(debug, "wasm log{}: httpCall() invalid timeout_milliseconds{}", log_prefix(),
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/context.cc:896]: Nit: please downgrade `debug` to `trace` (here and everywhere else).

**Review by PiotrSikora — COMMENTED** [review]
Agreed, this is a minimal risk change, but it's adding a ton of trace logs for debugging purposes, which isn't really suitable for a production software, IMHO. Also, it's adding those logs only for a single feature, and nothing else.

I don't mind if Envoy maintainers approve this and it's merged, but I won't be doing it myself. cc @lizan @mathetake

**Comment by esnible** [comment]
@PiotrSikora Fair enough.  I have greatly reduce the size of this PR.  I have removed all logging on the Wasm success path and reduced the logging on the WASM failure path from DEBUG to TRACE.  Does this make it palatable?

As an aside, how do you recommend users debug their WASM filters?  I am using Istio.  I put the .wasm binary in a ConfigMap, mount it on the Istio sidecar, and attach it using an EnvoyFilter.  When I call an ABI function and get a `WasmResult::BadArgument` I can log it from my filter but without a way to trace into the Envoy C++ code I am really stuck if the reason isn't obvious.  It is also easy to crash Envoy with accidental indirect infinite recursion in a buggy filter.  Is there a group looking into better sandbox and better filter debug experience?

**Comment by mathetake** [comment]
Actually, I suffered from this when I was initially developing my Go SDK, and I'd done the exactly the same thing in this PR in my local Envoy builds. So it makes sense for code alone.

But I agree with @PiotrSikora, and if I could, I would add these logging to all of other Wasm calls.

The point is that we should think about host calls of Proxy-Wasm as "system-calls" just like WASI, and WASI runtimes never emits this kind of debugging messages, but we never suffer from this kind of debugging issue with WASI since it has more descriptive and variety of [error codes](https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md#variants-1) returned by WASI "system-call"s.

In any way, we should update proxy-wasm's spec and define more kinds of error status for users and SDK devs to debug their programs easily.

File: source/extensions/common/wasm/context.cc [hunk 3]
```diff
@@ -902,6 +905,15 @@ WasmResult Context::httpCall(absl::string_view cluster, const Pairs& request_hea
...
       message->headers().Host() == nullptr) {
+    if (message->headers().Path() == nullptr) {
+      ENVOY_LOG(trace, "wasm log{}: path required", log_prefix());
+    }
+    if (message->headers().Method() == nullptr) {
+      ENVOY_LOG(trace, "wasm log{}: method required", log_prefix());
+    }
+    if (message->headers().Host() == nullptr) {
+      ENVOY_LOG(trace, "wasm log{}: host required", log_prefix());
+    }
```
Comment (lizan) [inline source/extensions/common/wasm/context.cc:916]: consolidate this to a one line log to something like:

ENVOY_LOG(trace, "wasm log{}: {}{}{}required", log_prefix(), message->headers().Path() == nullptr ? "path" : "", ...);

**Review by lizan — COMMENTED** [review]
> But I agree with @PiotrSikora, and if I could, I would add these logging to all of other Wasm calls as well.

+1, can you file an issue to track that?

> The point is that we should think about host calls of Proxy-Wasm as "system-calls" just like WASI, and WASI runtimes never emits this kind of debugging messages, but we never suffer from this kind of debugging issue with WASI since it has more descriptive and variety of error codes returned by WASI "system-call"s.

That's probably a separate topic about the ABI spec. Regards to debug messages for "system-call"s, we should be able to do `strace` equivalent for those "system-call"s.

**Comment by PiotrSikora** [comment]
> Regards to debug messages for "system-call"s, we should be able to do `strace` equivalent for those "system-call"s.

Agreed. We had that originally, but it was "lost" during migration to the [proxy-wasm-cpp-host](https://github.com/proxy-wasm/proxy-wasm-cpp-host) repo :(

**Comment by mathetake** [comment]
> That's probably a separate topic about the ABI spec. Regards to debug messages for "system-call"s, we should be able to do strace equivalent for those "system-call"s.

Totally agree +1, and let me take a look envoy-wasm repo and rescue "strace" in proxy-wasm

With regard to error code varieties, I will file an issue in proxy-wasm/spec repo

**Comment by mathetake** [comment]
Ah, I see what Lizan means by "strace equivalent" is this kind of trace logs in this PR. Got it.

**Comment by mathetake** [comment]
Can we you use this api and the host call to provide debugging functionality? I guess these are the ones @PiotrSikora mentioned as "lost" in the migration since `status_message_` is almost always empty string in the Envoy's current implementation
- https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/include/proxy-wasm/context_interface.h#L563-L567
- https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/include/proxy-wasm/exports.h#L63

**Comment by PiotrSikora** [comment]
> Can we use this api and the host call to provide debugging functionality? I guess these are the ones @PiotrSikora mentioned as "lost" in the migration since `status_message_` is almost always empty string in the Envoy's current implementation
> 
> * https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/include/proxy-wasm/context_interface.h#L563-L567
> * https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/include/proxy-wasm/exports.h#L63

No, what I meant is that we had a real trace messages for each call that crossed ABI boundary, including all parameters and return values, see:
* [source/extensions/common/wasm/v8/v8.cc#L597-L598](https://github.com/envoyproxy/envoy-wasm/blob/ef24be7f256f8e858e4d47f328963b55a9991ddd/source/extensions/common/wasm/v8/v8.cc#L597-L598),
* [source/extensions/common/wasm/v8/v8.cc#L604](https://github.com/envoyproxy/envoy-wasm/blob/ef24be7f256f8e858e4d47f328963b55a9991ddd/source/extensions/common/wasm/v8/v8.cc#L604),
* [source/extensions/common/wasm/v8/v8.cc#L658](https://github.com/envoyproxy/envoy-wasm/blob/ef24be7f256f8e858e4d47f328963b55a9991ddd/source/extensions/common/wasm/v8/v8.cc#L658),
* [source/extensions/common/wasm/v8/v8.cc#L667](https://github.com/envoyproxy/envoy-wasm/blob/ef24be7f256f8e858e4d47f328963b55a9991ddd/source/extensions/common/wasm/v8/v8.cc#L667).

You can still see those if you run Istio Proxy 1.6 with trace logs, e.g.:
```
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:626] [host->vm] proxy_on_tick(1)
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:595] [vm->host] env.proxy_get_current_time_nanoseconds(1048312)
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:525] setMemory(1048312, 8)
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:601] [vm<-host] env.proxy_get_current_time_nanoseconds return: 0
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:595] [vm->host] env.proxy_log(2, 1114184, 62)
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:516] getMemory(1114184, 62)
[info][wasm] [source/extensions/common/wasm/context.cc:1107] wasm log : It's 2020-12-07 06:50:04.304281 UTC, there is no lucky number.
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:601] [vm<-host] env.proxy_log return: 0
[trace][wasm] [source/extensions/common/wasm/v8/v8.cc:634] [host<-vm] proxy_on_tick return: void
```

**Comment by mathetake** [comment]
Looks fantastic! I really would like to retrieve these trace logs, and it's going to be fairly simple to get it implemented in proxy-wasm-cpp-host repo

**Comment by mathetake** [comment]
https://github.com/envoyproxy/envoy/pull/14433
https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/118

now that strace equivalent is implemented in proxy-wasm-cpp-host(https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/118), and Envoy's side integration is ready(https://github.com/envoyproxy/envoy/pull/14433)

**Comment by mathetake** [comment]
in addition to the strace I introduced above PR, maybe it's useful to implement `getStatus` ABI for more detailed messages (it's easy to implement by just setting `status_message_` to the ones like in this PR.

@PiotrSikora thoughts?

**Pull request closed without merging**